File: JAVASCRIPT/Add_Two_number.js


function ListNode(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
  
  /**
   * @param {ListNode} l1
   * @param {ListNode} l2
   * @return {ListNode}
   */
  var addTwoNumbers = function (l1, l2) {
    let carry = 0;
    let additionNode = new ListNode(0);
    let ansNode = additionNode;
  
    while (l1 || l2 || carry) {
      let sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry;
      carry = Math.floor(sum / 10);
      sum %= 10;
      ansNode.next = new ListNode(sum);
  
      l1 = l1 ? l1.next : null;
      l2 = l2 ? l2.next : null;
      ansNode = ansNode.next;
    }
  
    return additionNode.next;
  };

File: C++/Add_Two_Number.cpp

#include <bits/stdc++.h>
using namespace std;


struct ListNode {
     int val;
     ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
 };



class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* p1 = l1;
        ListNode* p2 = l2;
        ListNode* result = new ListNode(-1);
        ListNode* dummy = result;
        while(p1 || p2 || carry){
            if(!p1 && carry == 0){
                result->next = p2;
                break;
            }
            if(!p2 && carry == 0){
                result->next = p1;
                break;
            }
            int digit1 = p1 ? p1->val : 0;
            int digit2 = p2 ? p2->val : 0;
            int sum = digit1 + digit2 + carry;
            carry = sum / 10;
            sum %= 10;
            result->next = new ListNode(sum);
            if(p1)  p1 = p1->next;
            if(p2)  p2 = p2->next;
            result = result->next;
        }
        result = dummy->next;
        delete dummy;
        return result;
    }
};







